bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -299,23 +299,45 @@ void ReplicaImp::tryToSendPrePrepareMsg(bool batchingLogic) {
   PrePrepareMsg *pp =
       new PrePrepareMsg(config_.replicaId, curView, (primaryLastUsedSeqNum + 1), firstPath, primaryCombinedReqSize);
 
-  ClientRequestMsg *nextRequest = requestsQueueOfPrimary.front();
-  while (nextRequest != nullptr && nextRequest->size() <= pp->remainingSizeForRequests()) {
-    MDC_CID_PUT(GL, nextRequest->getCid());
-    if (clientsManager->noPendingAndRequestCanBecomePending(nextRequest->clientProxyId(),
-                                                            nextRequest->requestSeqNum())) {
-      pp->addRequest(nextRequest->body(), nextRequest->size());
-      clientsManager->addPendingRequest(nextRequest->clientProxyId(), nextRequest->requestSeqNum());
+  uint16_t initialStorageForRequests = pp->remainingSizeForRequests();
+
+  ClientRequestMsg *nextRequest = (!requestsQueueOfPrimary.empty() ? requestsQueueOfPrimary.front() : nullptr);
+
+  while (nextRequest != nullptr) {
+    if (nextRequest->size() <= pp->remainingSizeForRequests()) {
+      MDC_CID_PUT(GL, nextRequest->getCid());
+      if (clientsManager->noPendingAndRequestCanBecomePending(nextRequest->clientProxyId(),
+                                                              nextRequest->requestSeqNum())) {
+        pp->addRequest(nextRequest->body(), nextRequest->size());
+        clientsManager->addPendingRequest(nextRequest->clientProxyId(), nextRequest->requestSeqNum());
+      }
+      primaryCombinedReqSize -= nextRequest->size();
+    } else if (nextRequest->size() > initialStorageForRequests) {
```

Comment:
why not just `else` here? It would be equivalent to `else if nextRequest->size() > pp->remainingSizeForRequests()`. If I understand correctly, `remainingSizeForRequests()` decreases with every new request you add to the PP message, whereas `initialStorageForRequests` always holds the initial storage size.

Reply:
This bug I'm trying to fix here occur when a client send a request that is too big to fit the space in PP, the current implementation doesn't handle those requests and leads to failure of the replica. 
As Andrew mentioned in his first comment, "during batching requests aren't evenly sized they unlikely to perfectly fill the available space in a pre-prepare", so it may lead us to spamming the logs. I want to log just those request that are too big to be added to the initial PP space.

Reply:
Alright, makes sense, thanks for the clarification.